text/042-unify-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,305 @@
+  ```
+
+  But it seems consistent with the current philosophy of the language
+  to give many synonyms for doing the same thing and trusting the user
+  to produce readable scripts in the end.
+
+- The first part of the CEP will impose users of the existing `Theorem
+  ... with` syntax to add a `recursive` attribute to their
+  declarations.
+
+- The second part of the CEP will impose users of the `Theorem` /
+  `Defined` combo to add an `unsealed` attribute to their declarations
+  or to set some of the new flags.  Thanks to the flags, this porting
+  step should be pretty uninvasive.
+
+# Alternatives
+
+- We could keep the syntax separate and keep waiting for users asking
+  for this or that extension to implement it, but it is unlikely that
+  the syntax would ever converge, and it would remain inconsistent and
+  confusing.  The first chapter of the refactored reference manual
+  could rely on the unsatisfactory `Proof @term` syntax to avoid
+  bringing the tactics in.
+
+- We could implement this plan without the second part.  This would
+  not be a big problem, but it would leave a small inconsistency
+  between the way transparency is set when the body is provided or
+  not.  In this case, the `sealed` and `unsealed` attributes would
+  raise an error when the body is not provided.
+
+- We could implement this plan without providing the `sealed` and
+  `unsealed` attributes at all.  This means that when a body is
+  provided, `Theorem` and `Definition` would not be interchangeable
+  (but they already mostly are when in proof mode because the `Qed`
+  and `Defined` commands are not forced upon the user, so this would
+  not be consistent).
+
+- We could also decide to implement this plan but make `Definition`
+  and `Theorem` fully interchangeable.  A `Theorem` with a body would
+  be transparent by default, a `sealed` attribute could be used on
+  `Definition` or `Theorem` to make the body opaque.  In this case, we
+  would not implement the second part of the CEP either.
+
+# Unresolved questions
+
+- The current `Theorem ... with` syntax has been criticized for
+  allowing proofs that are rejected at `Qed` time (unless the
+  `Guarded` command is used).  The problem is exactly the same as when
+  using the `fix` or `cofix` tactics.  When using `Fixpoint` in proof
+  mode, it would seem that the `@fixannot` clause is ignored.  We
+  could look into making this clause being used and the use of tactic
+  becoming safer, but IMHO as a first step it suffices to properly
+  document this behavior and possibly warn when using the `@fixannot`
+  clause that it is ignored.
+
+- Another command that should be merged with `Definition` is
+  `Instance` (an `instance` attribute could be supported on
+  `Definition` and variants).  However, given that `Instance` itself
+  has several additional peculiarities (supports `{ }` syntax, accepts
+  auto-generating a name for the instance), this could also happen as
+  a later step.
```

Comment:
As discussed in https://github.com/coq/ceps/pull/42#discussion_r384727546, while fixing that issue, it'd be great to also address https://github.com/coq/coq/issues/7913 by allowing folded fixpoints, including mutually recursive ones — just run into that again.

That comment is now hidden — It'd be good to insert the link in the CEP itself.